Guides for Drupal and TYPO3 (#1553)
platformsh/platformsh-docs: docs/src/guides/typo3/deploy/customize.md

In order to make these variables, and any other you set in the future, easier to read and use within TYPO3 it's recommended that you install Platform.sh's [Configuration Reader library for PHP](https://github.com/platformsh/config-reader-php). It contains a set of common functions for interacting with the environment within a PHP application. 

```bash
$ composer require platformsh/config-reader
```

### Avoiding deadlock with the Local Page Error Handler

A default TYPO3 installation has a risk of deadlocks when run on low-worker PHP-FPM configurations. Specifically, TYPO3 handles 403 and 404 error pages by issuing a full HTTP request back to itself with no timeout, which can lead to process starvation. There are two required steps to avoid this problem.

First, you will need to install the [Local Page Error Handler](https://extensions.typo3.org/extension/pxa_lpeh/) extension for TYPO with the command:

```bash

> $ composer require pixelant/pxa-lpeh

Reviewer: Given that the user has a working TYPO3 instance already, you can enable the extension at this point too.

This is copied from the extension manual:

Enable the extension in the Admin tools > Extensions module or run `vendor/bin/typo3 extension:activate pxa_lpeh`

```

Second, you will need to add a timeout which will set an HTTP timeout of 3 seconds instead of the default several minutes to a new `public/typo3conf/PlatformshConfiguration.php`. You can see this line in context of the full file in the [configuration](#environment) section below.

```php
$GLOBALS['TYPO3_CONF_VARS']['HTTP']['timeout'] = 3;

> ```

Reviewer: Alert the user that a timeout of three seconds may be too little if the TYPO3 instance is supposed to perform external requests other than to itself.

If the instance will make long requests, for example synchronizing data with external APIs as a part of a TYPO3 Scheduler task, it is best to place these operations in workers.

You will still need to enable the `pixelant/pxa-lpeh` extension, which you will be able to do in the [post-install](/guides/typo3/deploy/deploy.md#post-install) steps later in this guide.

### (Optional) Introductory examples

The TYPO3 template for Platform.sh loads a number of example content pages provided by TYPO3 into the site when dependencies are installed pre-build. Their inclusion is completely optional, but a helpful choice if you are creating a new TYPO3 site from scratch using Composer.

```bash
$ composer require typo3/cms-introduction

> ```

Reviewer: This is not relevant for an existing site.

## Configuration

You're going to add and/or edit a few files described in the sections below, which will result in the following directory structure:

```bash
.
├── config
│   └── sites
│       └── main
│           └── config.yaml
...
├── public
│   └── typo3conf
│       ├── AdditionalConfiguration.php
│       └── PlatformshConfiguration.php
└── src
    ├── SetupConfiguration.yaml

>     └── SetupDatabase.yaml

Reviewer: If you're starting with an existing site, it's not so likely that you have these files. They're not a standard part of TYPO3.

```

### Setup

It's likely that you have a setup configuration file (`src/SetupConfiguration.yaml`) already present in your repository which manages your installation during build when running the `typo3 install:setup` command. On Platform.sh, it is recommended that that file  include at least the following commands.

> {{< github repo="platformsh-templates/typo3" file="src/SetupConfiguration.yaml" lang="yaml" >}}

Reviewer: I suggest removing this. It is required when installing TYPO3 from scratch.

### Site

> Similarly, you likely have a `config.yaml` file in your repo, althout it may not be in the `config/sites/main` subdirectory as shown above. This file contains settings for the error handling of your site, the languages it will support, and a few others. For the purposes of this guide, you will need to set the `base` attribute to an environment variable called `PLATFORM_ROUTES_MAIN`. 

Reviewer: ```suggestion
You will have to locate the site configuration file(s), `config.yaml`, in your repository's `config/sites/<SITEID>` subdirectories. For the purposes of this guide, you will need to set the `base` attribute to an environment variable called `PLATFORM_ROUTES_MAIN`. 
```

I think you'll want to show how to do this. The simplest version is line 1 in this file: https://github.com/t3kit/t3kit-starter/blob/master/.localconf/platformsh/config/sites/site2/config.yaml#L1

However, the best practice would be to define a baseVariant for the production environment, along the lines of this file: https://github.com/t3kit/t3kit-starter/blob/master/config/sites/t3kit/config.yaml#L2-L5

{{< github repo="platformsh-templates/typo3" file="config/sites/main/config.yaml" lang="yaml" >}}

You will define this environment variable in the next sections, but it's purpose is to retrieve the root domain (since you have not yet registered a domain name on the Platform.sh project, this will be a hashed placeholder domain generated from the environment) from the environment varable `PLATFORM_ROUTES`.

### Database

Next you will need a `src/SetupDatabase.yaml` file, which is used on the first installation of TYPO3 in the  dedploy hook to set up the database and the initial `admin` user.

> {{< github repo="platformsh-templates/typo3" file="src/SetupDatabase.yaml" lang="yaml" >}}

Reviewer: You can skip this. Instead, explain how to push an existing DB's content to Platform.sh.

The PlatformshConfiguration.php file should set all of the DB credentials automatically, so if that one is included correctly, it no more configuration should be necessary.

### Environment

> Finally, you can start using the Platform.sh Configuration Reader library to starting reading from the environment from within your application. In a `public/typo3confg/PlatformshConfiguration.php` file, you can use the library to:

Reviewer: ```suggestion
Finally, you can start using the Platform.sh Configuration Reader library to starting reading from the environment from within your application. In a `public/typo3conf/PlatformshConfiguration.php` file, you can use the library to:
```

- verify the deployment is occuring on a Platform.sh project (`if (!$platformConfig->isValidPlatform())`)
- verify that it is not run during build, when services are not yet available (`if ($platformConfig->inBuild())`)
- set the `PLATFORM_ROUTES_MAIN` environment variable used in `config/sites/main/config.yaml`)
- configure TYPO3's database (`TYPO3_CONF_VARS.DB.Connections.Default`) using credentials from the `database` relationship (`$databaseConfig = $platformConfig->credentials('database')`)
- configure TYPO3's `cacheConfigurations` to use Redis via your `rediscache` relationship
- Configure the HTTP timeout to 3 seconds to avoid the PHP-FPM-related deadlock described above in [Avoiding deadlock with the Local Page Error Handler](#avoiding-deadlock-with-the-local-page-error-handler).

{{< github repo="platformsh-templates/typo3" file="public/typo3conf/PlatformshConfiguration.php" lang="php" >}}

> Then include the `require_once()` function within your `public/typo3conf/AdditionalConfiguration.php` file to load the Platform.sh-specific configuration into the site.

Reviewer: I don't think it's been explained yet where to get this file from. Check that all files that are necessary are added to the filesystem at the beginning of the guide.
